### PR TITLE
[FEAT] Implement Blog Search and Filtering API #25

### PR DESCRIPTION
**Pull request by Ibukun-AL**
[FEAT] Implement Blog Search and Filtering API #25
### How should this be manually tested?

1. **Perform a search query:**
   - Make a request to the search endpoint with a valid search term.
   - The expected result is to receive a list of blog posts that match the search term.

2. **Apply filters:**
   - Make a request to the search endpoint with various filters (e.g., category, tags, date range).
   - The expected result is to receive a filtered list of blog posts based on the applied filters.

3. **Pagination:**
   - Make a request to the search endpoint with pagination parameters (e.g., page number, page size).
   - The expected result is to receive a paginated list of blog posts.

4. **Invalid search term:**
   - Make a request to the search endpoint with an invalid or empty search term.
   - The expected result is to receive an appropriate error response.

5. **Invalid filters:**
   - Make a request to the search endpoint with invalid filter values.
   - The expected result is to receive an appropriate error response.

### Checklist of what you did

- Implemented a search and filter endpoint for blog posts.
- Integrated server-side validation for search queries and filters.
- Added logging for search and filter operations.
- Implemented pagination support for the search results.
- Handled invalid search terms and filter values with proper error responses.
- Integrated with Typesense for search functionality.
- Ensured performance and scalability of the search feature.

### Reference Issue

[FEAT]: Blog Search and Filter API - [GitHub Issue #25](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/25)

### CODE REVIEW

File: src/main/java/hng_java_boilerplate/blog/controller/BlogController.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import hng_java_boilerplate.blog.service.BlogService;
+
+
+
+@RestController
+@RequestMapping("/api/v1/blogs")
+public class BlogController {
+
+    @Autowired
+    private BlogService blogService;
```
Comment (Fifetoyi) [inline src/main/java/hng_java_boilerplate/blog/controller/BlogController.java:30]: Use constructor injection with @RequiredArgsConstructor

File: src/main/java/hng_java_boilerplate/blog/service/BlogService.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+import hng_java_boilerplate.blog.dtos.BlogSearchResponse;
+import hng_java_boilerplate.blog.models.Blog;
+import hng_java_boilerplate.blog.repository.BlogRepository;
+
+@Service
+public class BlogService {
+
+    @Autowired
+    private BlogRepository blogRepository;
```
Comment (Fifetoyi) [inline src/main/java/hng_java_boilerplate/blog/service/BlogService.java:27]: Use constructor injection for all instance variables

File: src/test/java/hng_java_boilerplate/blog_test/BlogControllerTest.java
Comment (Fifetoyi) [inline src/test/java/hng_java_boilerplate/blog_test/BlogControllerTest.java:1]: No e2e tests available

**Comment by Am0du** [comment]
Make a proper pull request, guideline is on slack

**Comment by Ibukun-AL** [comment]
Please review again

**Pull request closed without merging**